File: libs/workspace/src/layer_tree_diff.cpp
// V14 tree diff — port of layer_tree_diff.py (see layer_tree_diff.hpp).
#include "pwb/workspace/layer_tree_diff.hpp"

#include <algorithm>
#include <map>
#include <set>

namespace pwb::workspace {

namespace {

// Patience LIS over an int sequence, returning chosen sequence indices.
std::vector<std::size_t> lis_chosen_positions(
    const std::vector<long long>& seq) {
    std::vector<long long> tails;
    std::vector<std::size_t> tails_pos;
    std::vector<std::int64_t> prev(seq.size(), -1);
    for (std::size_t idx = 0; idx < seq.size(); ++idx) {
        const long long value = seq[idx];
        auto pos_it = std::lower_bound(tails.begin(), tails.end(), value);
        const std::size_t pos = static_cast<std::size_t>(pos_it - tails.begin());
        if (pos == tails.size()) {
            tails.push_back(value);
            tails_pos.push_back(idx);
        } else {
            tails[pos] = value;
            tails_pos[pos] = idx;
        }
        prev[idx] = pos > 0 ? static_cast<std::int64_t>(tails_pos[pos - 1]) : -1;
    }
    std::vector<std::size_t> chosen;
    std::int64_t node =
        tails_pos.empty() ? -1 : static_cast<std::int64_t>(tails_pos.back());
    while (node >= 0) {
        chosen.push_back(static_cast<std::size_t>(node));
        node = prev[static_cast<std::size_t>(node)];
    }
    return chosen;
}

struct TreeIndex {
    // parent node id ("") -> child ids (in order). Only parents that have
    // children appear (diff iterates the union).
    std::map<std::string, std::vector<std::string>> children_of;
    // node id -> parent id.
    std::map<std::string, std::string> parents;
    // group id -> node.
    std::map<std::string, const TreeNode*> groups;
};

std::string node_id_of(const TreeNode& node) {
    return node.is_group ? node.group_id : node.layer_id;
}

void index_tree_into(const std::vector<TreeNode>& children,
                     const std::string& parent_id, TreeIndex& index) {
    for (const TreeNode& child : children) {
        const std::string node_id = node_id_of(child);
        if (child.is_group) index.groups[node_id] = &child;
        index.parents[node_id] = parent_id;
        index.children_of[parent_id].push_back(node_id);
        if (child.is_group) index_tree_into(child.children, node_id, index);
    }
}

TreeIndex index_tree(const LayerTreeSnapshot& snapshot) {
    TreeIndex index;
    index_tree_into(snapshot.children, "", index);
    return index;
}

// Whether each desired position belongs to the order-preserving
// "do not move" set (current<->desired common subsequence).
std::vector<bool> lcs_membership(const std::vector<std::string>& current_ids,
                                 const std::vector<std::string>& desired_ids) {
    std::map<std::string, std::vector<std::size_t>> current_positions;
    for (std::size_t idx = 0; idx < current_ids.size(); ++idx) {
        current_positions[current_ids[idx]].push_back(idx);
    }
    // Reduction: for each desired id in order, append its current
    // positions reversed (the classic patience mapping).
    std::vector<long long> seq;
    std::vector<std::size_t> des_index_of;
    for (std::size_t di = 0; di < desired_ids.size(); ++di) {
        auto it = current_positions.find(desired_ids[di]);
        if (it == current_positions.end()) continue;
        for (std::size_t ri = it->second.size(); ri-- > 0;) {
            seq.push_back(static_cast<long long>(it->second[ri]));
            des_index_of.push_back(di);
        }
    }
    const std::vector<std::size_t> chosen = lis_chosen_positions(seq);
    std::vector<bool> membership(desired_ids.size(), false);
    for (const std::size_t k : chosen) membership[des_index_of[k]] = true;
    return membership;
}

}  // namespace

std::vector<std::size_t> lcs_indices(const std::vector<std::string>& a,
                                     const std::vector<std::string>& b) {
    if (a == b) {
        std::vector<std::size_t> out(a.size());
        for (std::size_t i = 0; i < a.size(); ++i) out[i] = i;
        return out;
    }
    std::map<std::string, std::vector<std::size_t>> b_positions;
    for (std::size_t idx = 0; idx < b.size(); ++idx) {
        b_positions[b[idx]].push_back(idx);
    }
    std::vector<long long> seq;
    std::vector<std::size_t> a_index_of;
    for (std::size_t a_idx = 0; a_idx < a.size(); ++a_idx) {
        auto it = b_positions.find(a[a_idx]);
        if (it == b_positions.end()) continue;
        for (std::size_t ri = it->second.size(); ri-- > 0;) {
            seq.push_back(static_cast<long long>(it->second[ri]));
            a_index_of.push_back(a_idx);
        }
    }
    const std::vector<std::size_t> chosen = lis_chosen_positions(seq);
    std::set<std::size_t> keep;
    for (const std::size_t k : chosen) keep.insert(a_index_of[k]);
    return std::vector<std::size_t>(keep.begin(), keep.end());
}

TreeDiff diff_trees(const LayerTreeSnapshot& current,
                    const LayerTreeSnapshot& desired) {
    const TreeIndex cur = index_tree(current);
    const TreeIndex des = index_tree(desired);

    TreeDiff diff;

    // -- group set + renames + keyed states ----------------------------------
    for (const auto& [group_id, group] : des.groups) {
        auto cur_it = cur.groups.find(group_id);
        if (cur_it == cur.groups.end()) {
            GroupCreate create;
            create.group_id = group_id;
            create.name = group->name;
            auto parent_it = des.parents.find(group_id);
            create.parent = parent_it == des.parents.end()
                                ? std::string()
                                : parent_it->second;
            diff.group_creates.push_back(std::move(create));
        } else {
            const TreeNode* cur_group = cur_it->second;
            if (cur_group->name != group->name) {
                diff.group_renames.push_back(
                    GroupRename{group_id, group->name});
            }
            if (cur_group->visible != group->visible) {
                diff.group_states.push_back(
                    GroupStateSet{group_id, "visible", group->visible});
            }
            if (cur_group->expanded != group->expanded) {
                diff.group_states.push_back(
                    GroupStateSet{group_id, "expanded", group->expanded});
            }
            if (cur_group->locked != group->locked) {
                diff.group_states.push_back(
                    GroupStateSet{group_id, "locked", group->locked});
            }
        }
    }
    for (const auto& [group_id, group] : cur.groups) {
        (void)group;
        if (des.groups.find(group_id) == des.groups.end()) {
            diff.group_removes.push_back(GroupRemove{group_id});
        }
    }

    // Creates in topological order (parents before children): depth in
    // the desired tree approximates a topological order.
    auto depth_of = [&des](const std::string& group_id) {
        int depth = 0;
        std::string node = group_id;
        while (!node.empty()) {
            auto it = des.parents.find(node);
            if (it == des.parents.end()) break;
            node = it->second;
            ++depth;
        }
        return depth;
    };
    std::stable_sort(diff.group_creates.begin(), diff.group_creates.end(),
                     [&](const GroupCreate& a, const GroupCreate& b) {
                         return depth_of(a.group_id) < depth_of(b.group_id);
                     });

    // -- placement and order: LCS over the container union -------------------
    std::set<std::string> all_containers;
    for (const auto& entry : cur.children_of) all_containers.insert(entry.first);
    for (const auto& entry : des.children_of) all_containers.insert(entry.first);
    for (const std::string& container : all_containers) {
        auto cur_it = cur.children_of.find(container);
        auto des_it = des.children_of.find(container);
        std::vector<std::string> cur_ids =
            cur_it == cur.children_of.end() ? std::vector<std::string>{}
                                            : cur_it->second;
        std::vector<std::string> des_ids =
            des_it == des.children_of.end() ? std::vector<std::string>{}
                                            : des_it->second;
        const std::vector<bool> in_lcs = lcs_membership(cur_ids, des_ids);
        for (std::size_t index = 0; index < des_ids.size(); ++index) {
            if (in_lcs[index]) continue;
            const std::string& node_id = des_ids[index];
            auto parent_it = des.parents.find(node_id);
            const std::string parent =
                parent_it == des.parents.end() ? container : parent_it->second;
            if (des.groups.count(node_id) != 0) {
                diff.group_moves.push_back(
                    GroupMove{node_id, parent, static_cast<int>(index)});
            } else {
                diff.layer_moves.push_back(
                    LayerMove{node_id, parent, static_cast<int>(index)});
            }
        }
    }
    return diff;
}

}  // namespace pwb::workspace

File: legacy/python_reference/product/paleo_workbench/mapping_workspace/layer_tree_diff.py
"""V11 TreeDiff：期望树 vs 当前树的最小操作集（keyed LCS）。

设计（docs/development/qgis-cartography-runtime-v11/05-tree-diff.md）：
替换"发现顺序不同 → 全 children 重排"的粗放路径。每个容器内对子 id
序列做最长公共子序列（LCS）匹配——已在 LCS 中的节点不动；不在 LCS 中
的节点按目标位置逐个 move。组集合差分 → create/remove；重命名与
显隐/展开/锁为独立键控操作。

输出操作为冻结 dataclass（可序列化、可在测试中断言数量），由
``LayerGroupController._apply_tree`` 消费并映射为桥调用（native
transaction 窗口内批量执行）。

复杂度：O(Σ|children|·log) （LCS via patience），无 list.index 热路径。
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Sequence

from paleo_workbench.mapping_workspace.layer_tree import (
    GroupNode,
    LayerRef,
    LayerTreeSnapshot,
)

__all__ = [
    "GroupCreate",
    "GroupRemove",
    "GroupRename",
    "GroupMove",
    "LayerMove",
    "GroupStateSet",
    "TreeDiff",
    "diff_trees",
    "lcs_indices",
]


@dataclass(frozen=True)
class GroupCreate:
    group_id: str
    name: str
    parent: str  # "" = root


@dataclass(frozen=True)
class GroupRemove:
    group_id: str


@dataclass(frozen=True)
class GroupRename:
    group_id: str
    new_name: str


@dataclass(frozen=True)
class GroupMove:
    group_id: str
    new_parent: str
    new_index: int


@dataclass(frozen=True)
class LayerMove:
    layer_id: str
    new_parent: str  # "" = root
    new_index: int


@dataclass(frozen=True)
class GroupStateSet:
    """组状态键控设置（kind ∈ visible/expanded/locked）。"""

    group_id: str
    kind: str
    value: bool


@dataclass(frozen=True)
class TreeDiff:
    group_creates: tuple[GroupCreate, ...] = ()
    group_removes: tuple[GroupRemove, ...] = ()
    group_renames: tuple[GroupRename, ...] = ()
    group_moves: tuple[GroupMove, ...] = ()
    layer_moves: tuple[LayerMove, ...] = ()
    group_states: tuple[GroupStateSet, ...] = ()

    @property
    def is_empty(self) -> bool:
        return not (self.group_creates or self.group_removes or self.group_renames
                    or self.group_moves or self.layer_moves or self.group_states)

    def op_count(self) -> int:
        return (len(self.group_creates) + len(self.group_removes)
                + len(self.group_renames) + len(self.group_moves)
                + len(self.layer_moves) + len(self.group_states))


def lcs_indices(a: Sequence[str], b: Sequence[str]) -> list[int]:
    """LCS：返回 ``a`` 中属于公共子序列的下标（升序）。

    O(n log n)（patience LIS 归约）；等长同序序列返回全部下标。
    """
    if list(a) == list(b):
        return list(range(len(a)))
    b_positions: dict[str, list[int]] = {}
    for idx, value in enumerate(b):
        b_positions.setdefault(value, []).append(idx)
    seq: list[int] = []
    a_index_of: list[int] = []
    for a_idx, value in enumerate(a):
        for pos in reversed(b_positions.get(value, ())):
            seq.append(pos)
            a_index_of.append(a_idx)
    import bisect

    tails: list[int] = []
    tails_pos: list[int] = []
    prev: list[int] = [-1] * len(seq)
    for idx, value in enumerate(seq):
        pos = bisect.bisect_left(tails, value)
        if pos == len(tails):
            tails.append(value)
            tails_pos.append(idx)
        else:
            tails[pos] = value
            tails_pos[pos] = idx
        prev[idx] = tails_pos[pos - 1] if pos > 0 else -1
    chosen: set[int] = set()
    node = tails_pos[-1] if tails_pos else -1
    while node >= 0:
        chosen.add(node)
        node = prev[node]
    return sorted({a_index_of[k] for k in chosen})


def _index_tree(snapshot: LayerTreeSnapshot):
    """(children_of: parent_id -> [node dict], parents: id -> parent_id,
    groups: id -> GroupNode)"""
    children_of: dict[str, list] = {}
    parents: dict[str, str] = {}
    groups: dict[str, GroupNode] = {}

    def walk(children, parent_id):
        children_of.setdefault(parent_id, [])
        for child in children:
            if isinstance(child, GroupNode):
                node_id = child.group_id
                groups[node_id] = child
            else:
                node_id = child.layer_id
            parents[node_id] = parent_id
            children_of[parent_id].append(child)
            if isinstance(child, GroupNode):
                walk(child.children, node_id)

    walk(snapshot.children, "")
    return children_of, parents, groups


def _child_ids(children) -> list[str]:
    return [c.group_id if isinstance(c, GroupNode) else c.layer_id for c in children]


def diff_trees(
    current: LayerTreeSnapshot,
    desired: LayerTreeSnapshot,
) -> TreeDiff:
    """最小操作集：current（观察/已应用）→ desired（期望）。

    约定：
    * 组删除不级联删除图层（C++ removeGroupsExcept 上提子节点——操作
      语义与桥一致）；组创建按拓扑序输出（父先于子）。
    * move 只对不在容器 LCS 中的节点发出（含跨容器移动）。
    * visible/expanded/locked 只对两侧都存在的组做键控比较。
    """
    cur_children, cur_parents, cur_groups = _index_tree(current)
    des_children, des_parents, des_groups = _index_tree(desired)

    creates: list[GroupCreate] = []
    removes: list[GroupRemove] = []
    renames: list[GroupRename] = []
    group_moves: list[GroupMove] = []
    layer_moves: list[LayerMove] = []
    states: list[GroupStateSet] = []

    # -- 组集合 + 重命名 + 状态 -------------------------------------------------
    for group_id, group in des_groups.items():
        if group_id not in cur_groups:
            creates.append(GroupCreate(
                group_id=group_id, name=group.name,
                parent=des_parents.get(group_id, "")))
        else:
            cur = cur_groups[group_id]
            if cur.name != group.name:
                renames.append(GroupRename(group_id, group.name))
            if cur.visible != group.visible:
                states.append(GroupStateSet(group_id, "visible", group.visible))
            if cur.expanded != group.expanded:
                states.append(GroupStateSet(group_id, "expanded", group.expanded))
            if cur.locked != group.locked:
                states.append(GroupStateSet(group_id, "locked", group.locked))
    for group_id in cur_groups:
        if group_id not in des_groups:
            removes.append(GroupRemove(group_id))

    # 创建按拓扑序（父先于子）：desired 树深度序即拓扑序的近似——用
    # 递归深度排序保证父组 create 排在子组之前。
    def _depth_of(group_id: str) -> int:
        depth = 0
        node = group_id
        while node:
            node = des_parents.get(node, "")
            depth += 1
        return depth

    creates.sort(key=lambda c: _depth_of(c.group_id))

    # -- 放置与顺序：容器并集上 LCS --------------------------------------------
    all_containers = set(cur_children) | set(des_children)
    for container in sorted(all_containers):
        cur_ids = _child_ids(cur_children.get(container, ()))
        des_ids = _child_ids(des_children.get(container, ()))
        # LCS 在「current 序列 vs desired 序列」上求：desired 中属于 LCS
        # 的位置已保序就位，无需 move。
        des_in_lcs = _lcs_membership(cur_ids, des_ids)
        for index, node_id in enumerate(des_ids):
            if des_in_lcs[index]:
                continue
            parent = des_parents.get(node_id, container)
            if node_id in des_groups:
                group_moves.append(GroupMove(node_id, parent, index))
            else:
                layer_moves.append(LayerMove(node_id, parent, index))

    return TreeDiff(
        group_creates=tuple(creates),
        group_removes=tuple(removes),
        group_renames=tuple(renames),
        group_moves=tuple(group_moves),
        layer_moves=tuple(layer_moves),
        group_states=tuple(states),
    )


def _lcs_membership(current_ids: Sequence[str], desired_ids: Sequence[str]) -> list[bool]:
    """desired 各位置是否属于 current↔desired 的公共子序列（保序不动集）。"""
    current_positions: dict[str, list[int]] = {}
    for idx, node_id in enumerate(current_ids):
        current_positions.setdefault(node_id, []).append(idx)
    # 归约为「desired 中存在且 current 中的位置序列」的 LIS：
    seq: list[int] = []
    for node_id in desired_ids:
        for pos in reversed(current_positions.get(node_id, ())):
            seq.append(pos)
    # patience LIS over seq，映射回 desired 下标
    import bisect

    tails: list[int] = []
    tails_pos: list[int] = []
    prev: list[int] = [-1] * len(seq)
    for idx, value in enumerate(seq):
        pos = bisect.bisect_left(tails, value)
        if pos == len(tails):
            tails.append(value)
            tails_pos.append(idx)
        else:
            tails[pos] = value
            tails_pos[pos] = idx
        prev[idx] = tails_pos[pos - 1] if pos > 0 else -1
    chosen: set[int] = set()
    node = tails_pos[-1] if tails_pos else -1
    while node >= 0:
        chosen.add(node)
        node = prev[node]
    membership = [False] * len(desired_ids)
    # seq 的构造顺序 = desired 顺序×current 位置倒序；chosen 的 seq 下标
    # 对应唯一的 desired 下标 = seq 下标 // 1？——seq 元素按 desired 顺序
    # 排列，其下标即 (desired_index, current_pos) 对的线性序；直接把
    # chosen 的 seq 下标映射回 desired 下标：seq[k] 属于 desired[k']，
    # k' = k // count(node) 不成立（每 desired 元素可有多个 current 位）。
    # 修正：LIS 选中的每个 seq 元素携带其 desired 下标。
    des_index_of: list[int] = []
    for di, node_id in enumerate(desired_ids):
        for _ in current_positions.get(node_id, ()):
            des_index_of.append(di)
    for k in chosen:
        membership[des_index_of[k]] = True
    return membership
